addon/instance-initializers/prefetch.js

```diff
@@ -38,7 +41,15 @@ export function initialize(instance) {
       }
 
       // Run the prefetch hook if the route has one.
-      handlerInfo.handler._prefetched = handlerInfo.runSharedModelHook(transition, 'prefetch', [fullParams]);
+      if (!handlerInfo.handler && handlerInfo.handlerPromise) {
+        transition._handlerPromiseChain = transition._handlerPromiseChain.then(() => (
+          handlerInfo.handlerPromise.then((handler) => {
+            handler._prefetched = handlerInfo.runSharedModelHook(transition, 'prefetch', [fullParams]);
+          })
+        ));
+      } else {
+        handlerInfo.handler._prefetched = handlerInfo.runSharedModelHook(transition, 'prefetch', [fullParams]);
```

Comment:
So does the `'the prefetch hook is invoked for async handlers'` test not model a real-world scenario?